src/Kestrel.Transport.Sockets/Internal/SocketAwaitableEventArgs.cs

```diff
@@ -5,74 +5,137 @@
+
+            // Use UserToken to carry the continuation state around
+            UserToken = state;
+            Action<object> awaitableState = Interlocked.CompareExchange(ref _completion, continuation, null);
 
-            if (SocketError != SocketError.Success)
+            if (ReferenceEquals(awaitableState, _callbackCompleted))
+            {
+#if NETCOREAPP2_1
+                ThreadPool.QueueUserWorkItem(continuation, state, preferLocal: false);
+#else
+                Task.Factory.StartNew(continuation, state);
+#endif
+            }
+            else if (awaitableState != null)
             {
-                ThrowSocketException(SocketError);
+                throw new InvalidOperationException("Multiple continuations registered!");
             }
 
-            return BytesTransferred;
+        }
+
+        public int GetResult(short token)
+        {
+            SocketError error = SocketError;
+            int bytes = BytesTransferred;
+
+            Reset();
+
+            if (error != SocketError.Success)
+            {
+                ThrowSocketException(error);
+            }
+
+            return bytes;
 
             void ThrowSocketException(SocketError e)
             {
                 throw new SocketException((int)e);
             }
         }
 
-        public void OnCompleted(Action continuation)
+        public ValueTask<int> ReceiveAsync(Socket socket)
         {
-            if (ReferenceEquals(_callback, _callbackCompleted) ||
-                ReferenceEquals(Interlocked.CompareExchange(ref _callback, continuation, null), _callbackCompleted))
+            if (socket.ReceiveAsync(this))
             {
-                Task.Run(continuation);
+                return new ValueTask<int>(this, 0);
             }
+
+            return CompleteSocketOperation();
         }
 
-        public void UnsafeOnCompleted(Action continuation)
+        /// <summary>Initiates a send operation on the associated socket.</summary>
+        /// <returns>This instance.</returns>
+        public ValueTask<int> SendAsync(Socket socket)
         {
-            OnCompleted(continuation);
+            if (socket.SendAsync(this))
+            {
+                return new ValueTask<int>(this, 0);
+            }
+
+            return CompleteSocketOperation();
         }
 
-        public void Complete()
+        private void Reset()
         {
-            OnCompleted(this);
+            Volatile.Write(ref _completion, null);
         }
 
-        protected override void OnCompleted(SocketAsyncEventArgs _)
+        private ValueTask<int> CompleteSocketOperation()
         {
-            var continuation = Interlocked.Exchange(ref _callback, _callbackCompleted);
+            int bytesTransferred = BytesTransferred;
+            SocketError error = SocketError;
 
-            if (continuation != null)
-            {
-                _ioScheduler.Schedule(state => ((Action)state)(), continuation);
-            }
+            Reset();
```

Comment:
Why call Reset() here? _completion must already be null, right? Which now that I think about it means GetStatus is wrong after sync completions. Again, I'm not sure how important that is.

Reply:
Sync completions don't use this code path. Unless I'm missing something...

Reply:
??? This is *only* called when ReceiveAsync/SendAsync return false, right?

I guess it's confusing because of the name. That's why I recommend renaming this to GetCompletedValueTask().